### PR TITLE
[ISSUE #70] Retained message based on raft state machine

### PR DESCRIPTION
**Pull request by YYYYWD**

### CODE REVIEW

**Comment by YYYYWD** [comment]
#70

File: meta/src/test/java/org/apache/rocketmq/mqtt/meta/raft/RetainedMsgClient.java [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+import org.apache.rocketmq.mqtt.common.model.consistency.Response;
+import org.apache.rocketmq.mqtt.common.model.consistency.WriteRequest;
+import org.apache.rocketmq.mqtt.common.util.TopicUtils;
+import org.apache.rocketmq.mqtt.meta.raft.processor.Constants;
+
+import java.util.HashMap;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+
+public class RetainedMsgClient {
```
Comment (DongyuanPan) [inline meta/src/test/java/org/apache/rocketmq/mqtt/meta/raft/RetainedMsgClient.java:44]: Please use unit test format

File: mqtt-common/src/main/java/org/apache/rocketmq/mqtt/common/util/MessageUtil.java [hunk 2]
```diff
@@ -39,22 +41,28 @@
 public class MessageUtil {
     public static final ByteBufAllocator ALLOCATOR = new UnpooledByteBufAllocator(false);
 
-    public static MqttPublishMessage toMqttMessage(String topicName, byte[] body, int qos, int mqttId) {
+
+
+    public static final String EMPTYSTRING = "%@!@%";
```
Comment (DongyuanPan) [inline mqtt-common/src/main/java/org/apache/rocketmq/mqtt/common/util/MessageUtil.java:46]: Use special notation, such as /n /r

**Comment by ShannonDing** [comment]
link #118

**Comment by ShannonDing** [comment]
the GID separator has been changed from "=" to "%",
rebase the latest develop branch and merge the PR #129 .

**Review by ShannonDing — CHANGES_REQUESTED** [review]
the GID separator has been changed from "=" to "%",
rebase the latest develop branch and merge the PR https://github.com/apache/rocketmq-mqtt/pull/129 .

**Comment by YYYYWD** [comment]
> the GID separator has been changed from "=" to "%", rebase the latest develop branch and merge the PR #129 .

Using % can cause unpredictable errors, so use "-" instead

**Comment by DongyuanPan** [comment]
@YYYYWD Y Using % can cause unpredictable errors, so use "-" instead. it can make a new pr to develop

File: meta/src/main/java/org/apache/rocketmq/mqtt/meta/raft/processor/Constants.java [hunk 3]
```diff
@@ -20,5 +20,7 @@
 public class Constants {
     public static final String COUNTER = "counter";
 
+    public static final String RETAINEDMSG = "retainedmsg";
```
Comment (DongyuanPan) [inline meta/src/main/java/org/apache/rocketmq/mqtt/meta/raft/processor/Constants.java:23]: retainedmsg -> retainedMsg

File: meta/src/main/java/org/apache/rocketmq/mqtt/meta/raft/processor/RetainedMsgStateProcess.java [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+
+    private final AtomicLong value = new AtomicLong(0);
+
+    private final HashMap<String, String> retainedMsgMap = new HashMap<>();  //key:topic value:retained msg
+
+    private final HashMap<String, Trie<String, String>> retainedMsgTopicTrie = new HashMap<>();  //key:firstTopic value:retained topic Trie
+
+    protected final ReentrantReadWriteLock lock = new ReentrantReadWriteLock();
+
+    private static final int LIMIT_RETAINED_MESSAGE_COUNT = 100;
```
Comment (DongyuanPan) [inline meta/src/main/java/org/apache/rocketmq/mqtt/meta/raft/processor/RetainedMsgStateProcess.java:49]: LIMIT_RETAINED_MESSAGE_COUNT should be config in service.conf

File: meta/src/main/java/org/apache/rocketmq/mqtt/meta/raft/processor/RetainedMsgStateProcess.java [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+    private SnapshotOperation snapshotOperation;
+
+
+    @Override
+    public Response onReadRequest(ReadRequest request) {
+        try {
+            String topic = request.getExtDataMap().get("topic");
+            String flag = request.getExtDataMap().get("flag");
```
Comment (DongyuanPan) [inline meta/src/main/java/org/apache/rocketmq/mqtt/meta/raft/processor/RetainedMsgStateProcess.java:57]: flag can be replace by type in ReadRequest

File: meta/src/main/java/org/apache/rocketmq/mqtt/meta/raft/processor/RetainedMsgStateProcess.java [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.function.BiConsumer;
+
+public class RetainedMsgStateProcess extends StateProcessor {
+
+    private final AtomicLong value = new AtomicLong(0);
+
+    private final HashMap<String, String> retainedMsgMap = new HashMap<>();  //key:topic value:retained msg
```
Comment (DongyuanPan) [inline meta/src/main/java/org/apache/rocketmq/mqtt/meta/raft/processor/RetainedMsgStateProcess.java:43]: Does the state machine need to care about message content?  Or use byte array storage directly, reducing serialization and deserialization

**Comment by DongyuanPan** [comment]
Does the state machine need to care about message content? Or use byte array storage directly, reducing serialization and deserialization. Trie tree retainedMsgTopicTrie  in state machine also has the same problem

**Comment by DongyuanPan** [comment]
To facilitate troubleshooting, you need to add error logs

File: meta/src/main/java/org/apache/rocketmq/mqtt/meta/raft/processor/RetainedMsgStateProcess.java (hunk 6 above)
Comment (DongyuanPan) [inline meta/src/main/java/org/apache/rocketmq/mqtt/meta/raft/processor/RetainedMsgStateProcess.java:43]: There may be concurrency issues

File: meta/src/main/java/org/apache/rocketmq/mqtt/meta/raft/processor/RetainedMsgStateProcess.java [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+
+    private final HashMap<String, Trie<String, String>> retainedMsgTopicTrie = new HashMap<>();  //key:firstTopic value:retained topic Trie
```
Comment (DongyuanPan) [inline meta/src/main/java/org/apache/rocketmq/mqtt/meta/raft/processor/RetainedMsgStateProcess.java:45]: There may be concurrency issues

File: meta/src/main/java/org/apache/rocketmq/mqtt/meta/raft/processor/RetainedMsgStateProcess.java [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+                .setErrMsg(e.getMessage())
+                .build();
+        }
+
+
+    }
+
+    @Override
+    public SnapshotOperation loadSnapshotOperate() {
+        snapshotOperation = new CounterSnapshotOperation(lock);
```
Comment (DongyuanPan) [inline meta/src/main/java/org/apache/rocketmq/mqtt/meta/raft/processor/RetainedMsgStateProcess.java:145]: You need to implement business-appropriate snapshot operations.

File: mqtt-ds/src/main/java/org/apache/rocketmq/mqtt/ds/meta/RetainedPersistManagerImpl.java [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+import java.util.Set;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+
+import java.util.concurrent.ScheduledThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+
+
+public class RetainedPersistManagerImpl implements RetainedPersistManager {
```
Comment (DongyuanPan) [inline mqtt-ds/src/main/java/org/apache/rocketmq/mqtt/ds/meta/RetainedPersistManagerImpl.java:47]: Tires should no longer be cached locally, and local TRIE should not be used to match topics, so there will still be caching delays.  It should be matched directly in the state machine

File: meta/src/main/java/org/apache/rocketmq/mqtt/meta/raft/processor/RetainedMsgStateProcess.java [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+            if (operation.equals("topic")) {    //return retained msg
+                byte[] msgBytes = retainedMsgMap.get(topic);
+                return Response.newBuilder()
+                    .setSuccess(true)
+                    .setData(ByteString.copyFrom(msgBytes))
+                    .build();
+            } else { //return retain msgs of matched Topic
+                if (!retainedMsgTopicTrie.containsKey(firstTopic)) {
+                    Trie<String, String> newTrie = new Trie<>();
+                    retainedMsgTopicTrie.put(firstTopic, newTrie);
```
Comment (DongyuanPan) [inline meta/src/main/java/org/apache/rocketmq/mqtt/meta/raft/processor/RetainedMsgStateProcess.java:70]: Judging that there is no such topic, after initializing the trie tree, you can return directly

File: mqtt-ds/src/main/java/org/apache/rocketmq/mqtt/ds/upstream/processor/PublishProcessor.java [hunk 11]
```diff
@@ -51,23 +57,49 @@ public class PublishProcessor implements UpstreamProcessor {
...
+
+        if (mqttPublishMessage.fixedHeader().isRetain()) {
+            //Change the retained flag of message that will send MQ is 0
+            mqttPublishMessage = MessageUtil.removeRetainedFlag(mqttPublishMessage);
+            //Keep the retained flag of message that will store meta
+            Message metaMessage = MessageUtil.toMessage(retainedMqttPublishMessage);
+            metaMessage.setMsgId(msgId);
+            metaMessage.setBornTimestamp(bornTime);
+            metaMessage.setEmpty(isEmpty);
+            CompletableFuture<Boolean> booleanCompletableFuture = retainedPersistManager.storeRetainedMessage(TopicUtils.normalizeTopic(metaMessage.getOriginTopic()), metaMessage);
```
Comment (DongyuanPan) [inline mqtt-ds/src/main/java/org/apache/rocketmq/mqtt/ds/upstream/processor/PublishProcessor.java:91]: if store retain message error, send message action will failed. it should be notify cs and user

**Comment by DongyuanPan** [comment]
Please adjust the code style, remove extra blank lines, and add spaces around certain symbols

File: meta/src/test/java/org/apache/rocketmq/mqtt/meta/raft/RetainedMsgClientTest.java [hunk 12]
```diff
@@ -0,0 +1,251 @@
...
+                public Executor executor() {
+                    return null;
+                }
+            }, 5000);
+        } catch (InterruptedException | RemotingException e) {
+            throw new RuntimeException(e);
+        }
+
+        future.whenComplete(((message, throwable) -> {
+            Mockito.verify(message, null);
```
Comment (DongyuanPan) [inline meta/src/test/java/org/apache/rocketmq/mqtt/meta/raft/RetainedMsgClientTest.java:195]: Compare whether the content of the message is stored

File: mqtt-common/src/main/java/org/apache/rocketmq/mqtt/common/util/TrieUtil.java [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+package org.apache.rocketmq.mqtt.common.util;
+
+import org.apache.rocketmq.mqtt.common.model.Trie;
+
+import java.util.HashSet;
+import java.util.Set;
+
+public class TrieUtil {
+
+    public static void mergeKvToLocal(Trie<String, String> localTrie, Trie<String, String> kvTrie) {
```
Comment (DongyuanPan) [inline mqtt-common/src/main/java/org/apache/rocketmq/mqtt/common/util/TrieUtil.java:28]: Do you still need this class?

File: mqtt-common/src/test/java/org/apache/rocketmq/mqtt/common/test/model/TestTrie.java [hunk 14]
```diff
@@ -91,10 +91,54 @@ public void test() {
...
+        //test 'getNode' by 'k/r/#'
+        String wcTopic = "k/r/#/";
+        Set<String> wcFilterSet = new HashSet<>(Arrays.asList("k/r/b/", "k/r/a/c/", "k/r/a/","k/r/c/","k/r/a/d/"));
+        Set<String> allPath = reverseTrie.getAllPath(wcTopic);
+        Assert.assertEquals(wcFilterSet,allPath);
+
+        wcTopic = "k/+/#/";
+        wcFilterSet = new HashSet<>(Arrays.asList("k/r/b/", "k/r/a/c/", "k/r/a/", "k/r/c/", "k/r/a/d/", "k/a/b/c/r/"));
+        allPath = reverseTrie.getAllPath(wcTopic);
+        Assert.assertEquals(wcFilterSet,allPath);
```
Comment (DongyuanPan) [inline mqtt-common/src/test/java/org/apache/rocketmq/mqtt/common/test/model/TestTrie.java:119]: wcFilterSet,allPath   , followed by a space

File: mqtt-common/src/test/java/org/apache/rocketmq/mqtt/common/test/util/TestMessageUtil.java [hunk 15]
```diff
@@ -69,7 +74,7 @@ public void Before() {
 
     @Test
     public void TestToMqttMessage() {
-        Assert.assertEquals(mqttPublishMessage.toString(), MessageUtil.toMqttMessage(topicName, messageBody.getBytes(), qos, mqttId).toString());
+        Assert.assertEquals(mqttPublishMessage.toString(), MessageUtil.toMqttMessage(topicName, messageBody.getBytes(), qos, mqttId,false).toString());
```
Comment (DongyuanPan) [inline mqtt-common/src/test/java/org/apache/rocketmq/mqtt/common/test/util/TestMessageUtil.java:77]: format code

File: mqtt-common/src/test/java/org/apache/rocketmq/mqtt/common/test/util/TestTrieUtil.java [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+import org.apache.rocketmq.mqtt.common.util.TrieUtil;
+import org.junit.Assert;
+import org.junit.Test;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+
+public class TestTrieUtil {
```
Comment (DongyuanPan) [inline mqtt-common/src/test/java/org/apache/rocketmq/mqtt/common/test/util/TestTrieUtil.java:32]: If TrieUtil is no longer needed, please delete this class too

Comment (DongyuanPan) [reply mqtt-common/src/test/java/org/apache/rocketmq/mqtt/common/test/util/TestTrieUtil.java:32]: Or move somewhere inside the state machine?

File: mqtt-cs/src/main/java/org/apache/rocketmq/mqtt/cs/protocol/mqtt/handler/MqttSubscribeHandler.java [hunk 17]
```diff
@@ -128,4 +160,66 @@ private MqttSubAckMessage getResponse(MqttSubscribeMessage mqttSubscribeMessage)
         return mqttSubAckMessage;
     }
 
+
+    @SuppressWarnings("checkstyle:Indentation")
```
Comment (DongyuanPan) [inline mqtt-cs/src/main/java/org/apache/rocketmq/mqtt/cs/protocol/mqtt/handler/MqttSubscribeHandler.java:164]: This should be deleted and the code should be formatted

File: mqtt-cs/src/main/java/org/apache/rocketmq/mqtt/cs/protocol/mqtt/handler/MqttSubscribeHandler.java [hunk 18]
```diff
@@ -128,4 +160,66 @@ private MqttSubAckMessage getResponse(MqttSubscribeMessage mqttSubscribeMessage)
...
+            });
+        }
+
+        for (Subscription subscription : wildcardTopics) {
+
+            CompletableFuture<ArrayList<String>> future = retainedPersistManager.getMsgsFromTrie(subscription);
+            future.whenComplete((msgsList,throwable) -> {
+                ArrayList<Message> results = new ArrayList<>();
+                for (String strMsg:msgsList) {
+                    results.add(JSON.parseObject(strMsg,Message.class));
```
Comment (DongyuanPan) [inline mqtt-cs/src/main/java/org/apache/rocketmq/mqtt/cs/protocol/mqtt/handler/MqttSubscribeHandler.java:196]: results.add(JSON.parseObject(strMsg,Message.class));    Can it be optimized here？

File: mqtt-ds/src/main/java/org/apache/rocketmq/mqtt/ds/retain/RetainedMsgClient.java [hunk 19]
```diff
@@ -0,0 +1,210 @@
...
+    public static void SetRetainedMsg(String topic, Message msg, CompletableFuture<Boolean> future) throws RemotingException, InterruptedException {
+
+        HashMap<String, String> option = new HashMap<>();
+        option.put("topic", topic);
+        option.put("firstTopic", msg.getFirstTopic());
+        option.put("isEmpty", String.valueOf(msg.isEmpty()));
+
+        logger.debug("SetRetainedMsg option:" + option);
+
+        final WriteRequest request = WriteRequest.newBuilder().setGroup(GROUPNAME + GROUP_SEQ_NUM_SPLIT + "0").setData(ByteString.copyFrom(JSON.toJSONBytes(msg, SerializerFeature.WriteClassName))).putAllExtData(option).build();
```
Comment (DongyuanPan) [inline mqtt-ds/src/main/java/org/apache/rocketmq/mqtt/ds/retain/RetainedMsgClient.java:106]: FYI  https://stackoverflow.com/questions/28545401/java-json-protobuf-back-conversion

File: mqtt-ds/src/main/java/org/apache/rocketmq/mqtt/ds/retain/RetainedMsgClient.java [hunk 20]
```diff
@@ -0,0 +1,210 @@
...
+                    Response rsp = (Response) result;
+                    if (!rsp.getSuccess()) {
+                        logger.info("GetRetainedMsg failed. {}", rsp.getErrMsg());
+                        return;
+                    }
+                    if (rsp.getData().toStringUtf8().equals("null")) {  //this topic doesn't exist retained msg
+                        return;
+                    }
+                    String strMsg = rsp.getData().toStringUtf8();
+                    Message message = JSON.parseObject(strMsg, Message.class);
```
Comment (DongyuanPan) [inline mqtt-ds/src/main/java/org/apache/rocketmq/mqtt/ds/retain/RetainedMsgClient.java:196]: Is there a better way for Convert Java object to protobuf or parseObject from protobuf？

File: mqtt-common/src/main/proto/storeMessage.proto [hunk 21]
```diff
@@ -0,0 +1,39 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+syntax = "proto3";
+
+option java_multiple_files = true;
+option java_package = "org.apache.rocketmq.mqtt.common.model.consistency";
+
+message StoreMessage {
```
Comment (DongyuanPan) [inline mqtt-common/src/main/proto/storeMessage.proto:23]: mqtt-common/src/main/proto/storeMessage.proto and mqtt-common/src/main/proto/request.proto should be merged and rename file name

**Review by DongyuanPan — APPROVED** [review]
GOOD JOB

**Comment by DongyuanPan** [comment]
@ShannonDing @pingww it can be merged

File: mqtt-cs/src/main/java/org/apache/rocketmq/mqtt/cs/protocol/mqtt/handler/MqttSubscribeHandler.java [hunk 22]
```diff
@@ -128,4 +159,63 @@ private MqttSubAckMessage getResponse(MqttSubscribeMessage mqttSubscribeMessage)
...
+                        return;
+                    }
+                    _sendMessage(session, clientId, subscription, msg);
+                }
+            });
+
+        }
+    }
+
+    private void _sendMessage(Session session, String clientId, Subscription subscription, Message message) {
```
Comment (pingww) [inline mqtt-cs/src/main/java/org/apache/rocketmq/mqtt/cs/protocol/mqtt/handler/MqttSubscribeHandler.java:203]: The send logic should converge to the push module

File: mqtt-cs/src/main/java/org/apache/rocketmq/mqtt/cs/protocol/mqtt/handler/MqttSubscribeHandler.java [hunk 23]
```diff
@@ -128,4 +159,63 @@ private MqttSubAckMessage getResponse(MqttSubscribeMessage mqttSubscribeMessage)
...
+            return;
+        }
+
+        int mqttId = mqttMsgId.nextId(clientId);
+        int qos = min(subscription.getQos(), message.qos());
+        if (qos == 0) {
+            pushAction.write(session, message, mqttId, 0, subscription);
+            pushAction.rollNextByAck(session, mqttId);
+        } else {
+            retryDriver.mountPublish(mqttId, message, subscription.getQos(), ChannelInfo.getId(session.getChannel()), subscription);
```
Comment (pingww) [inline mqtt-cs/src/main/java/org/apache/rocketmq/mqtt/cs/protocol/mqtt/handler/MqttSubscribeHandler.java:216]: New subscribers are all subscribed from the maximum site, RETAIN messages cannot be retried, if wildcard subscriptions are not in the queue. Whether a retry test has been performed？

**Comment by pingww** [comment]
to resolve conflicts

**Review by pingww — CHANGES_REQUESTED** [review]

**Review by pingww — CHANGES_REQUESTED** [review]

File: meta/src/main/java/org/apache/rocketmq/mqtt/meta/raft/processor/CounterStateProcessor.java [hunk 24]
```diff
@@ -34,9 +34,9 @@
 public class CounterStateProcessor extends StateProcessor {
```
Comment (pingww) [inline meta/src/main/java/org/apache/rocketmq/mqtt/meta/raft/processor/CounterStateProcessor.java:34]: What does this class do

File: mqtt-ds/src/main/java/org/apache/rocketmq/mqtt/ds/retain/RetainedMsgClient.java [hunk 25]
```diff
@@ -0,0 +1,218 @@
...
+        raftRpcFactory.registerProtobufSerializer(WriteRequest.class.getName(), WriteRequest.getDefaultInstance());
+        raftRpcFactory.registerProtobufSerializer(ReadRequest.class.getName(), ReadRequest.getDefaultInstance());
+        raftRpcFactory.registerProtobufSerializer(Response.class.getName(), Response.getDefaultInstance());
+
+        MarshallerRegistry registry = raftRpcFactory.getMarshallerRegistry();
+        registry.registerResponseInstance(WriteRequest.class.getName(), Response.getDefaultInstance());
+        registry.registerResponseInstance(ReadRequest.class.getName(), Response.getDefaultInstance());
+    }
+
+    public static void SetRetainedMsg(String topic, Message msg, CompletableFuture<Boolean> future) throws RemotingException, InterruptedException {
```
Comment (pingww) [inline mqtt-ds/src/main/java/org/apache/rocketmq/mqtt/ds/retain/RetainedMsgClient.java:97]: Method name prefixes are recommended in lower case

File: mqtt-cs/src/main/java/org/apache/rocketmq/mqtt/cs/protocol/mqtt/handler/MqttSubscribeHandler.java (hunk 23 above)
Comment (pingww) [reply mqtt-cs/src/main/java/org/apache/rocketmq/mqtt/cs/protocol/mqtt/handler/MqttSubscribeHandler.java:216]: That is, the retain message is not in the normal queue, which may affect the retry logic

File: meta/src/main/java/org/apache/rocketmq/mqtt/meta/raft/processor/CounterStateProcessor.java (hunk 24 above)
Comment (YYYYWD) [reply meta/src/main/java/org/apache/rocketmq/mqtt/meta/raft/processor/CounterStateProcessor.java:34]: This is a demo that state machine how to work

File: mqtt-cs/src/main/java/org/apache/rocketmq/mqtt/cs/protocol/mqtt/handler/MqttSubscribeHandler.java (hunk 23 above)
Comment (YYYYWD) [reply mqtt-cs/src/main/java/org/apache/rocketmq/mqtt/cs/protocol/mqtt/handler/MqttSubscribeHandler.java]: Through experimentation and log observations, everything works fine in this case.

File: mqtt-cs/src/main/java/org/apache/rocketmq/mqtt/cs/protocol/mqtt/handler/MqttSubscribeHandler.java (hunk 22 above)
Comment (YYYYWD) [reply mqtt-cs/src/main/java/org/apache/rocketmq/mqtt/cs/protocol/mqtt/handler/MqttSubscribeHandler.java]: Fix

**Pull request merged**